install.bat

```diff
@@ -471,6 +471,35 @@ local function get_visual_studio_directory()
 	return nil
 end
 
+local function get_visual_studio_directory_from_vswhere()
+	assert(type(vars.LUA_RUNTIME)=="string", "requires vars.LUA_RUNTIME to be set before calling this function.")
+	local major, minor = vars.LUA_RUNTIME:match('VCRUNTIME(%d+)(%d)$')
+	if not major or tonumber(major) < 14 then
+		print(S[[    Cannot auto-detect Visual Studio 2017 or higher from $LUA_RUNTIME]])
+		return nil
+	end
+	local vswhere_dir = os.getenv('PROGRAMFILES(X86)').."\\Microsoft Visual Studio\\Installer\\vswhere.exe"
+	if not exists(vswhere_dir) then
+		return nil
+	end
+	local f = io.popen('"'..vswhere_dir..'" -products * -property installationPath')
+	if not f then return nil, "failed to run vswhere " end
+	local vsdir = nil
+	while true do
+		local l = f:read()

```

Comment:
Should check for errors from `:read`

Reply:
What is the change you are propose here? `l` is already checked for nil on the following line of `read`.

Reply:
```lua
local l, err = f:read()
if not l then
    f:close()
    return nil, err
end
```

Reply:
I'm afraid this might not be the correct logic. What I was trying to do, is going through the output of `vswhere` and set `vsdir` to the last line of the output (the higher version of Visual Studio appears at last). And when `l` becomes `nil`, it means eof is reached and the while loop should stop. It does not make sense to return nil just because of eof.

Reply:
Ah right, it doesn't always mean EOF; it can also mean that you're missing permissions to read the output for some reason; or some other weird error: it's useful to return that error message.
So it needs to me more like:
```lua
local l, err = f:read()
if not l then
    if err then
        f:close()
        return nil, err
    else
        break
    end
end
```
Also note that you need to `:close()` after the loop in the general case.

Reply:
Fixed.